### PR TITLE
Cp2

### PR DESCRIPTION
**Pull request by chuckha**

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -76,3 +76,33 @@
...
+
+[[constraint]]
+  branch = "release-1.9"
+  name = "k8s.io/kubernetes"
+
+# To use reference package:
+#   vendor/k8s.io/kubernetes/pkg/util/parsers/parsers.go:36:16: undefined: reference.ParseNormalizedNamed
+[[override]]
+  name = "github.com/docker/distribution"
+  branch = "master"
```
Comment (chuckha) [inline Gopkg.toml]: this is fixed in the next commit

File: Gopkg.toml [hunk 2]
```diff
@@ -76,3 +76,33 @@
...
+
+# github.com/heptio/sonobuoy/vendor/k8s.io/kubernetes/pkg/credentialprovider
+# vendor/k8s.io/kubernetes/pkg/credentialprovider/keyring.go:61:2: undefined: "github.com/heptio/sonobuoy/vendor/github.com/docker/docker/api/types".AuthConfig
+[[override]]
+  name = "github.com/docker/docker"
+  branch = "master"
```
Comment (chuckha) [inline Gopkg.toml]: this is fixed in the next commit

File: cmd/sonobuoy/app/cp.go [hunk 3]
```diff
@@ -41,11 +43,32 @@ func init() {
 	RootCmd.AddCommand(cmd)
 }
 
+// copy from namespace heptio-sonobuoy from pod sonobuoy dir /tmp/sonobuoy to local ./archive file
```
Comment (chuckha) [inline cmd/sonobuoy/app/cp.go]: delete these

File: cmd/sonobuoy/app/operations/cp.go [hunk 4]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/kubectl/cmd"
 )
 
-func CopyResults(path string) error {
-	// pretty straight forwards
-	return errors.New("not implemented")
+// FileSpec defines what and where
```
Comment (chuckha) [inline cmd/sonobuoy/app/operations/cp.go]: use better words

File: cmd/sonobuoy/app/operations/cp.go [hunk 5]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+			fmt.Printf("error running: %v", err)
+		}
+
+	}()
+	prefix := strings.TrimLeft(src.File, "/")
+	prefix = path.Clean(prefix)
+	return untarAll(reader, dest.File, prefix)
+}
+
+func untarAll(reader io.Reader, destFile, prefix string) error {
```
Comment (chuckha) [inline cmd/sonobuoy/app/operations/cp.go]: this is copy pasted from k8s/k8s

File: cmd/sonobuoy/app/operations/cp.go [hunk 6]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+	if entrySeq == -1 {
+		//if no file was copied
+		errInfo := fmt.Sprintf("error: %s no such file or directory", prefix)
+		return errors.New(errInfo)
+	}
+	return nil
+}
+
+// dirExists checks if a path exists and is a directory.
+func dirExists(path string) (bool, error) {
```
Comment (chuckha) [inline cmd/sonobuoy/app/operations/cp.go]: this is copy pasted from k8s/k8s

File: cmd/sonobuoy/app/operations/cp.go (hunk 4 above)
Comment (liztio) [reply cmd/sonobuoy/app/operations/cp.go]: > Filespec defines what file whe are copying and where in the cluster we are copying it from
maybe?

**Comment by timothysc** [comment]
Is this pruned?  please say no...

File: cmd/sonobuoy/app/operations/cp.go [hunk 7]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+	}
+
+	go func() {
+		options.Config = config
+		options.PodClient = clientset.Core()
+
+		defer outStream.Close()
+
+		if err := options.Validate(); err != nil {
+			fmt.Println("error validating")
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/cp.go]: could we somehow bubble these errors up so we exit with nonzero if they don't succeed?

Comment (liztio) [reply cmd/sonobuoy/app/operations/cp.go]: these should probably go to stderr too

File: cmd/sonobuoy/app/operations/cp.go [hunk 8]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+	}()
+	prefix := strings.TrimLeft(src.File, "/")
+	prefix = path.Clean(prefix)
+	return untarAll(reader, dest.File, prefix)
+}
+
+func untarAll(reader io.Reader, destFile, prefix string) error {
+	entrySeq := -1
+
+	// TODO: use compression here?
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/cp.go]: this comment probably makes more sense on the compressing end of things than here?

File: cmd/sonobuoy/app/operations/cp.go [hunk 9]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+			if err != io.EOF {
+				return err
+			}
+			break
+		}
+		entrySeq++
+		mode := header.FileInfo().Mode()
+		outFileName := path.Join(destFile, header.Name[len(prefix):])
+		baseName := path.Dir(outFileName)
+		if err := os.MkdirAll(baseName, 0755); err != nil {
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/cp.go]: tar should be transmitting the mask, may as well use it (or comment as to why we're not)

File: cmd/sonobuoy/app/operations/cp.go [hunk 10]
```diff
@@ -17,10 +17,138 @@ limitations under the License.
...
+			}
+			if err := outFile.Close(); err != nil {
+				return err
+			}
+		}
+	}
+
+	if entrySeq == -1 {
+		//if no file was copied
+		errInfo := fmt.Sprintf("error: %s no such file or directory", prefix)
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/cp.go]: stderr?

File: cmd/sonobuoy/app/operations/cp.go (hunk 5 above)
Comment (liztio) [reply cmd/sonobuoy/app/operations/cp.go]: maybe note where they were copied from in a comment? might be useful to future archaeologists

**Comment by chuckha** [comment]
@timothysc this is unfortunately pruned.

File: Gopkg.toml [hunk 11]
```diff
@@ -76,3 +76,33 @@
...
   name = "k8s.io/client-go"
+
+[[constraint]]
+  branch = "release-1.9"
+  name = "k8s.io/kubernetes"
+
+# To use reference package:
+#   vendor/k8s.io/kubernetes/pkg/util/parsers/parsers.go:36:16: undefined: reference.ParseNormalizedNamed
+[[override]]
+  name = "github.com/docker/distribution"
```
Comment (timothysc) [inline Gopkg.toml]: Dear god, why do we need this?

File: cmd/sonobuoy/app/operations/cp.go (hunk 10 above)
Comment (chuckha) [reply cmd/sonobuoy/app/operations/cp.go]: `Sprintf` is returning a string, so this should be ok

File: cmd/sonobuoy/app/operations/cp.go (hunk 9 above)
Comment (chuckha) [reply cmd/sonobuoy/app/operations/cp.go]: i think the code is right unless i'm misunderstanding. It's making a dir locally (./archive) that may not exist yet and then putting the contents of the thing it's copying down into it

File: cmd/sonobuoy/app/operations/cp.go (hunk 10 above)
Comment (liztio) [reply cmd/sonobuoy/app/operations/cp.go]: oh, sure is. In this case you could use `fmt.Errorf` and combine this line with the next one.

**Comment by timothysc** [comment]
Closing this PR, b/c of other one.

**Pull request closed without merging**